File: www/src/lib/demo-message.ts
/** Keep in lockstep with tests/QikLog.Core.Tests/DemoMessage.cs and www/api/demo-send.js */
export const MAX_LENGTH = 120;

export function sanitize(raw: string | null | undefined): string {
  if (!raw) return '';
  let out = '';
  for (const ch of raw.trim()) {
    const code = ch.charCodeAt(0);
    if (code <= 0x1f || code === 0x7f) continue;
    out += ch;
    if (out.length >= MAX_LENGTH) break;
  }
  return out;
}

export function escapeJson(value: string): string {
  return value
    .replaceAll('\\', '\\\\')
    .replaceAll('"', '\\"')
    .replaceAll('\n', '\\n')
    .replaceAll('\r', '\\r')
    .replaceAll('\t', '\\t');
}

export function escapeCSharp(value: string): string {
  return escapeJson(value);
}

export function escapeJavaScript(value: string): string {
  return escapeJson(value);
}

export function escapeFor(kind: 'json' | 'csharp' | 'js', value: string): string {
  if (kind === 'csharp') return escapeCSharp(value);
  if (kind === 'js') return escapeJavaScript(value);
  return escapeJson(value);
}

File: tests/QikLog.Core.Tests/DemoMessage.cs
using System.Text;

namespace QikLog.Core.Tests;

/// <summary>
/// Spec for the landing demo-send sanitizer and per-language sample escaping.
/// Keep in lockstep with <c>www/src/lib/demo-message.ts</c> and <c>www/api/demo-send.js</c>.
/// </summary>
internal static class DemoMessage
{
    public const int MaxLength = 120;

    public static string Sanitize(string? raw)
    {
        if (string.IsNullOrEmpty(raw))
            return "";

        var builder = new StringBuilder(raw.Length);
        foreach (var ch in raw.Trim())
        {
            if (ch is >= '\u0000' and <= '\u001F' or '\u007F')
                continue;
            builder.Append(ch);
            if (builder.Length >= MaxLength)
                break;
        }

        return builder.ToString();
    }

    public static string EscapeJson(string value)
    {
        var builder = new StringBuilder(value.Length + 8);
        foreach (var ch in value)
        {
            switch (ch)
            {
                case '\\': builder.Append("\\\\"); break;
                case '"': builder.Append("\\\""); break;
                case '\n': builder.Append("\\n"); break;
                case '\r': builder.Append("\\r"); break;
                case '\t': builder.Append("\\t"); break;
                default: builder.Append(ch); break;
            }
        }

        return builder.ToString();
    }

    public static string EscapeCSharp(string value)
    {
        var builder = new StringBuilder(value.Length + 8);
        foreach (var ch in value)
        {
            switch (ch)
            {
                case '\\': builder.Append("\\\\"); break;
                case '"': builder.Append("\\\""); break;
                case '\n': builder.Append("\\n"); break;
                case '\r': builder.Append("\\r"); break;
                case '\t': builder.Append("\\t"); break;
                default: builder.Append(ch); break;
            }
        }

        return builder.ToString();
    }

    public static string EscapeJavaScript(string value) => EscapeJson(value);
}
